core/xmodel/cross_query_cache.go

```diff
@@ -0,0 +1,269 @@
+package xmodel
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"os"
+	"sync"
+	"time"
+
+	log "github.com/xuperchain/log15"
+	crypto_client "github.com/xuperchain/xuperchain/core/crypto/client"
+	"github.com/xuperchain/xuperchain/core/crypto/hash"
+	"github.com/xuperchain/xuperchain/core/pb"
+)
+
+const (
+	timeOut = 6
+)
+
+// CrossQueryCache cross query struct
+type CrossQueryCache struct {
+	lg               log.Logger
+	crossQueryCaches []*pb.CrossQueryInfo
+	crossQueryIdx    int
+	isPenetrate      bool
+}
+
+type queryRes struct {
+	*pb.CrossQueryResponse
+	*pb.SignatureInfo
+}
+
+// NewCrossQueryCache return CrossQuery instance while preexec
+func NewCrossQueryCache() *CrossQueryCache {
+	lg := log.New("module", "cross_query_cache")
+	lg.SetHandler(log.StreamHandler(os.Stderr, log.LogfmtFormat()))
+	return &CrossQueryCache{
+		lg:          lg,
+		isPenetrate: true,
+	}
+}
+
+// NewCrossQueryCacheWirthData return CrossQuery instance while posttx
+func NewCrossQueryCacheWirthData(crossQueries []*pb.CrossQueryInfo) *CrossQueryCache {
+	lg := log.New("module", "cross_query_cache")
+	lg.SetHandler(log.StreamHandler(os.Stderr, log.LogfmtFormat()))
+	return &CrossQueryCache{
+		lg:               lg,
+		crossQueryCaches: crossQueries,
+		isPenetrate:      false,
+	}
+}
+
+// CrossQuery query contract from otherchain
+func (cqc *CrossQueryCache) CrossQuery(
+	crossQueryRequest *pb.CrossQueryRequest,
+	queryMeta *pb.CrossQueryMeta) (*pb.ContractResponse, error) {
+	cqc.lg.Info("Receive CrossQuery", "crossQueryRequest", crossQueryRequest, "queryMeta", queryMeta)
+	if !isQueryMetaValid(queryMeta) {
+		return nil, fmt.Errorf("isQueryParamValid check failed")
+	}
+	// Call endorsor for responce
+	if cqc.isPenetrate {
```

Comment:
I think xmodel is a data abstract layer, is it better to implement endorser call in xbridge call?

Reply:
Now `xmodel` is a data model that can distinguish process `preexec` and `posttx`. Like `model cache`, `utxo cache`, `cross query cache`.